Throw error when exporting non-declaration

kaicataldo opened the pull request:
<!--- 
Before making a PR please make sure to read our contributing guidelines 
https://github.com/babel/babylon/blob/master/CONTRIBUTING.md
-->

| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| Breaking change?  | yes
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | yes
| Tests added/pass? | yes
| Fixed tickets     | #238
| License           | MIT

<!-- Describe your changes below in as much detail as possible -->
Currently, Babylon will parse any export that starts with a keyword as a declaration. These changes will throw when the statement following the `export` keyword is not a declaration.

DrewML commented on src/parser/statement.js:
```diff
@@ -861,8 +861,22 @@ pp.parseExport = function (node) {
-  return this.parseStatement(true);
+  const statement = this.parseStatement(true);
+  if (!this.isDeclaration(statement)) {
+    this.unexpected(statement.start);
```
Since we know what the problem is here (an `export` declaration that doesn't have a valid export declaration or export clause) maybe we can use `this.raise` and hint to the user what the problem was?

kaicataldo replied:
After discussing a bit with @hzoo, I changed it to an ahead of time check rather than checking the export statement post-parse. This actually changes the error message in some cases, so wanted to see what do you think. 

If we want a more descriptive error message I could leave the original conditional and do the `shouldParseExportDeclaration` check (maybe now called `isDeclaration`?) inside that block, throwing a more descriptive error if `isDeclaration` returns `false`.

My only concern with the error message is that the intention of the code is. If Babylon is trying to parse `export typeof foo`, can we surmise they intended to export a declaration? Or did you mean a more generic `Not a valid export` type of message? Let me know if I'm missing something!

kaicataldo commented:
Hmmm, actually, going to explore checking token values ahead of time rather than the node type afterwards

kaicataldo commented:
Updated

kaicataldo commented:
Sorry, one last update - added return value flow type and moved the `shouldParseExportDeclaration` method back to where it was before I removed the other methods

danez commented:
Nice thanks
